utils/pattern/src/implementations.rs

```diff
@@ -0,0 +1,26 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use zerovec::{maps::ZeroMapKV, ule::VarULE, VarZeroSlice, VarZeroVec, ZeroVecError};
+
+use crate::{Pattern, SinglePlaceholder, SinglePlaceholderPattern};
+
+impl<'a> ZeroMapKV<'a> for Pattern<SinglePlaceholder, str> {
+    type Container = VarZeroVec<'a, Pattern<SinglePlaceholder, str>>;
+    type Slice = VarZeroSlice<Pattern<SinglePlaceholder, str>>;
+    type GetType = Pattern<SinglePlaceholder, str>;
+    type OwnedType = Box<Pattern<SinglePlaceholder, str>>;
+}
+
+unsafe impl VarULE for Pattern<SinglePlaceholder, str> {
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), ZeroVecError> {
+        SinglePlaceholderPattern::try_from_bytes_store(bytes)
+            .map_err(|_| ZeroVecError::VarZeroVecFormatError)?;
+        Ok(())
+    }
+
+    unsafe fn from_byte_slice_unchecked(bytes: &[u8]) -> &Self {
+        &SinglePlaceholderPattern::try_from_bytes_store(bytes).unwrap()
+    }
```

Comment:
@sffc , it seems we are having issue with `try_from_bytes_store`,  

See the compiler error:

```
error[E0308]: mismatched types
  --> utils/pattern/src/implementations.rs:24:9
   |
23 |     unsafe fn from_byte_slice_unchecked(bytes: &[u8]) -> &Self {
   |                                                          ----- expected `&frontend::Pattern<SinglePlaceholder, str>` because of return type
24 |         &SinglePlaceholderPattern::try_from_bytes_store(bytes).unwrap()
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&Pattern<SinglePlaceholder, str>`, found `&Pattern<SinglePlaceholder, &str>`
   |
   = note: expected reference `&frontend::Pattern<_, _>`
              found reference `&frontend::Pattern<_, &_>`
```

Reply:
I made a fix here: https://github.com/unicode-org/icu4x/pull/5072

Reply:
That's not the right fix for this. You're trying to go from `&str` to `&Pattern<SinglePlaceholder, str>`. With the code in #5072, `SinglePlaceholderPattern::try_from_bytes_store(bytes)` has type `Pattern<SinglePlaceholder, String>` (`<str as ToOwned>::Owned` is `String`), so it allocates into an owned string. You're then returning a reference to this, which is a dangling reference and will not compile.

To go from `&str` to `&Pattern<_, str>` you will need to transmute, as that's the only way to create unsized types.

Reply:
done